### DISCUSSION TITLE
Peerjs on Samsung TV (Tizen OS)

### DISCUSSION THREAD

**Discussion by Robloche** [peers/peerjs#1140] [General]
Hello,

Has anyone tried to use peerjs in a web app running on a Samsung TV?

I wrote a short code that can act as receiver or sender (you can check it here: https://ntgrdb2.azureedge.net/index.html).
When I run it on 2 different browsers, I can indeed send data from the sender to the receiver.
But when the receiver runs on the TV, I get errors.

The connection to the signaling server is OK and the receiver seems to connect to the sender (see message `Connected to: xxx`), but then it fails shortly afterward:
```
PeerJS:  Server message received: {src: "3e4215e7-b320-455e-9a09-2080d66bb305", dst: "3eed59f2-5e8c-4259-92e0-7c1fa3359b33", payload: {…}, type: "OFFER"}
PeerJS:  Creating RTCPeerConnection.
PeerJS:  Listening for ICE candidates.
PeerJS:  Listening for data channel
PeerJS:  Listening for remote stream
PeerJS:  Setting remote description RTCSessionDescription {type: "offer", sdp: "v=0↵o=- 8267987733839877376 2 IN IP4 127.0.0.1↵s…:0↵a=sctp-port:5000↵a=max-message-size:262144↵"}
PeerJS:  add connection data:dc_62ajhcpxxhn to peerId:3e4215e7-b320-455e-9a09-2080d66bb305
[on.connection]
Connected to: 3e4215e7-b320-455e-9a09-2080d66bb305
ERROR PeerJS:  Error: (OperationError) Failed to execute 'setRemoteDescription' on 'RTCPeerConnection': Failed to parse SessionDescription. a=extmap-allow-mixed Expects at least 2 fields.
[on.error] 
PeerJS:  Failed to setRemoteDescription,  (OperationError) Failed to execute 'setRemoteDescription' on 'RTCPeerConnection': Failed to parse SessionDescription. a=extmap-allow-mixed Expects at least 2 fields.
PeerJS:  Server message received: {src: "3e4215e7-b320-455e-9a09-2080d66bb305", dst: "3eed59f2-5e8c-4259-92e0-7c1fa3359b33", payload: {…}, type: "CANDIDATE"}
PeerJS:  handleCandidate: {candidate: "candidate:1254127035 1 udp 2113937151 b69b7048-28b…typ host generation 0 ufrag 5JR7 network-cost 999", sdpMLineIndex: 0, sdpMid: "0", usernameFragment: "5JR7"}
PeerJS:  Server message received: {src: "3e4215e7-b320-455e-9a09-2080d66bb305", dst: "3eed59f2-5e8c-4259-92e0-7c1fa3359b33", payload: {…}, type: "CANDIDATE"}
PeerJS:  handleCandidate: {candidate: "candidate:1553511523 1 udp 1677729535 37.58.174.23… rport 0 generation 0 ufrag 5JR7 network-cost 999", sdpMLineIndex: 0, sdpMid: "0", usernameFragment: "5JR7"}
ERROR PeerJS:  Error: (OperationError) Failed to execute 'addIceCandidate' on 'RTCPeerConnection': Error processing ICE candidate
[on.error] 
PeerJS:  Failed to handleCandidate,  (OperationError) Failed to execute 'addIceCandidate' on 'RTCPeerConnection': Error processing ICE candidate
```

Any help appreciated.

**Comment by jonasgloning** [comment]
Hey @Robloche,
I haven’t tried it (and I don’t have access to a smart TV), but I think I have seen this error in some old chrome versions. Looking at the [Samsung TV Web Engine versions](https://developer.samsung.com/smarttv/develop/specifications/web-engine-specifications.html) I would expect it to work on Tizen 6 or 6.5 and above.
What OS version does your TV run?

**Reply by Robloche** [reply]
Thanks for your answer.

The TV I'm using runs Tizen 5.5, and that would make Chromium v69. Not very recent...
Do you know what's causing this issue inside Peerjs?
Will I get the same problem with other WebRTC libraries?
Otherwise I'll simply get a new TV...

As a side note, my app works perfectly in the Tizen TV simulator.

**Reply by EricEisaman** [reply]
@Robloche 
Maybe TMI for this thread but:

To force a Samsung TV OS update to a version newer than Tizen 5.5, you can try the following steps:

- Check for updates: Go to the Settings menu on your Samsung TV and navigate to the "Support" or "Software Update" section. Look for an option to check for updates manually. If available, select it to see if there are any new updates for your TV.
- Reset the TV: If checking for updates doesn't work, you can try resetting your TV to factory settings. This may prompt the TV to check for updates again and potentially install a newer version of the OS. To reset your TV, go to the Settings menu, navigate to the "General" or "System" section, and look for an option to reset or initialize the TV.
- Update using a USB drive: If the above methods don't work, you can try updating the TV's OS using a USB drive. Samsung provides firmware updates for some of its TVs on their website. You can download the latest firmware for your specific TV model from the Samsung support website and follow the instructions provided to update the TV using a USB drive.

**Reply by Robloche** [reply]
My experience with half a dozen of Samsung TVs is that they tend to stick to a major version of the OS.
Like, if a TV runs Tizen 2.3, it only accepts updates < v3.
But I will of course try what you suggest.
Thanks a lot!

**Reply by Robloche** [reply]
So, I tried everything:

- No way to manually check updates. And the automatic way, does not found Firmware above v5.5.
- Reset everything did not caused a firmware update.
- I tried 4 different and more recent firmwares via a USB key but every time, the TV told me that no update was available.

So, I guess I'm left with only one choice: asking my company for a new TV.

**Comment by Robloche** [comment]
Also, I see the following function in the generated `peerjs.js`:
```js
function $5151b78094cea2ba$export$a57d114344295149(window, browserDetails) {
    /* remove a=extmap-allow-mixed for webrtc.org < M71 */ if (!window.RTCPeerConnection) return;
    if (browserDetails.browser === "chrome" && browserDetails.version >= 71) return;
    if (browserDetails.browser === "safari" && browserDetails.version >= 605) return;
    const nativeSRD = window.RTCPeerConnection.prototype.setRemoteDescription;
    window.RTCPeerConnection.prototype.setRemoteDescription = function setRemoteDescription(desc) {
        if (desc && desc.sdp && desc.sdp.indexOf("\na=extmap-allow-mixed") !== -1) {
            const sdp = desc.sdp.split("\n").filter((line)=>{
                return line.trim() !== "a=extmap-allow-mixed";
            }).join("\n");
            // Safari enforces read-only-ness of RTCSessionDescription fields.
            if (window.RTCSessionDescription && desc instanceof window.RTCSessionDescription) arguments[0] = new window.RTCSessionDescription({
                type: desc.type,
                sdp: sdp
            });
            else desc.sdp = sdp;
        }
        return nativeSRD.apply(this, arguments);
    };
}
```

Isn't it supposed to act as a workaround for a bug existing in Chrome versions < 71?

**Comment by jonasgloning** [comment]
I just tried Chrome 69 with your site. Pressing the _Connect Server_ button resulted in these errors in the console:
<img width="582" alt="image" src="https://github.com/peers/peerjs/assets/34194370/2e77b9b4-e425-4f95-8c59-bfa6cb2c25af">
 (I’m sorry it’s low quality screenshot, I can’t copy from Browserstack)
 
Do you used exactly this site on your TV?

**Reply by Robloche** [reply]
No, I had to rewrite parts of it because Tizen does not support modern JS.
